### DISCUSSION TITLE
Using enums to drive navigation (from the Modern SwiftUI series) Question - how to derive the destination binding.

### DISCUSSION THREAD

**Discussion by Dominic263** [pointfreeco/swift-composable-architecture#1868] [Q&A]
Hi, 

I was re-watching the series on best practices in modelling navigation in an app and I came across a small issue. I know the Scrumdinger app is demoed without using TCA and its just fine but I am not quite sure what to do to drive modal sheets using the API provided by SwiftUINavigation in an app modeled with TCA. So for example, to create a simple sheet that pops up when a button is tapped I modelled a State which had a destination variable shown below, then used that to drive the navigation but I am just not sure how to derive the destination binding for the unwrapping: argument. here is a piece of code to explain what I'm thinking. Thanks for your help. 

```
public struct Feature: ReducerProtocol {
    public init() {}
    
    public struct State: Equatable {
        public var destination: Destination?
        public init(
            destination: Destination? = nil
        ) {

            self.destination = destination
        }
    }
    
    // Destination enum - navigate to different views and modals within the view
    public enum Destination: Equatable {
        // Sheet to add a new sprint
        case addSprint(Sprint)
    }
    
    public Action: Equatable { 
         case destinationChanged(Destination) 
   }


// ..... reducer and some other irrelevant  code down here 

}
```

Then in the view 

```
.sheet(
                    unwrapping: viewstore.binding(
                        get: \.destination,
                        send:  Feature.Action.destinationChanged
                    ),
                    case: /HomeFeature.Destination.addSprint
                ) { sprint in
                    PopOverView(store: self.store)
                        
                }
```

**Comment by Dominic263** [comment]
UPDATE:  I was reading one of your blogs on [SwiftUI bindings](https://www.pointfree.co/blog/posts/63-the-composable-architecture-%EF%B8%8F-swiftui-bindings) to see if I was doing something wrong and I saw a new way to derive bindings of parts of State using @BindableState property wrapper around the property then using the BindableAction protocol on the Action before using the value as a binding in the View like so: 

```
.sheet(
                    unwrapping: viewstore.binding(\.$dest)
                   
                    ,
                    case: /HomeFeature.Destination.addSprint
                ) { sprint in
                    PopOverView(store: self.store)
                        .background(Color.teal.opacity(0.5))
                }
                
```
This seems to work, but I was wondering if anyone knows why the old method isn't working in my case and what the proper way of driving modals is. thanks

**Comment by Dominic263** [comment]
Anyone figured out how to use the new SwiftUINavigation tools in TCA yet ? I know I could wait for next week's episode
 on sheets and popovers but I am curious how others have implemented this. Thanks

**Reply by mbrandonw** [reply]
Hi @Dominic263, the SwiftUINavigation library isn't really meant to be used with TCA. It's a separate thing that mostly just helps people write vanilla SwiftUI apps with better domain modeling tools.

We will be releasing the TCA navigation beta next week, so you won't have to wait much longer. 🙂

**Reply by Dominic263** [reply]
Oh gotcha I thought it fit perfectly well into TCA and was experimenting with some tools but I kept running into issues! Thanks Brandon, Can't wait.

**Comment by jaroslavas** [comment]
I tried to do something similar to what author did, and almost everything works, except reducer:

```
struct Feature: ReducerProtocol {

  enum Destination: Equatable {
    case menu(ConfirmationDialogState<Action>)
    case anotherScreen(AnotherFeature.State)
  }

  struct State: Equatable {
    var destination: Destination?
    // Many other state variables
  }

  enum Action: Equatable {
    case anotherFeatureActions(AnotherFeature.Action)
    // other actions
  }
 
  var body: some ReducerProtocol<State, Action> {
    Reduce { state, action in 
      // ...
    }

    // The question is how to add a child reducer here? 
//    .ifLet(\.destination, action: /Action.anotherFeatureActions) {
//        AnotherFeature()
//    }

  }

}

```

I just can't come up with how to scope it to extract child state from `Feature.destination.anotherScreen(let childState)`. Any ideas?

**Reply by mbrandonw** [reply]
This isn't easily achievable with the latest release of TCA, but we do have an [ongoing navigation-beta](https://github.com/pointfreeco/swift-composable-architecture/discussions/1944) that you can target if you want to live on the edge. Check out the case studies and tests to see how to run navigation off of a destination enum.

These tools will be making it into an official release of the library in the coming weeks.
